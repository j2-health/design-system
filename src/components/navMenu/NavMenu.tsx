import * as React from 'react'
import { ItemType } from 'antd/es/menu/interface'
import cx from 'classnames'
import { Menu } from '../menu'

import Logo from '../../svgs/j2-logo.svg?react'
import Wordmark from '../../svgs/j2-wordmark.svg?react'
import ExpandIcon from '../../svgs/menu-sidebar-expand.svg?react'
import CollapseIcon from '../../svgs/menu-sidebar-collapse.svg?react'
import { Button } from '../button'

import styles from './NavMenu.module.css'
import { Tooltip } from '../tooltip'

type Props = {
  href?: string
  items: ItemType[]
  headerSlot?: React.ReactNode
  defaultOpenKeys?: string[]
  defaultSelectedKeys?: string[]
  footerItems: ItemType[]
}

export const NavMenu = ({
  href = '/',
  items,
  headerSlot,
  defaultOpenKeys,
  defaultSelectedKeys,
  footerItems,
}: Props) => {
  const [isCollapsed, setIsCollapsed] = React.useState(false)
  const [selectedKey, setSelectedKey] = React.useState<string>(
    defaultSelectedKeys?.[0] || ''
  )
  const [isToggleTooltipOpen, setToggleTooltipOpen] = React.useState(false)

  const toggleCollapse = React.useCallback(() => {
    setIsCollapsed(!isCollapsed)
    setToggleTooltipOpen(false)
  }, [isCollapsed])

  return (
    <div
      className={cx(
        styles.navMenuContainer,
        isCollapsed && styles.collapsed,
        'bg-white'
      )}
    >
      <div className={cx('flex', 'flex-col', 'h-full')}>
        <div className={cx('py-5', isCollapsed ? 'px-4' : 'px-3')}>
          <div
            className={cx(
              'flex',
              'items-center',
              isCollapsed ? 'justify-center' : 'justify-between',
              styles.headerContainer,
              'min-h-9'
            )}
          >
            <a
              href={href}
              className={cx('flex', 'items-center', 'gap-2', styles.logo)}
            >
              <Logo />
              {!isCollapsed && <Wordmark />}
            </a>
            <Tooltip
              title={isCollapsed ? 'Expand Menu' : 'Collapse Menu'}
              placement="right"
              arrow={false}
              onOpenChange={(shouldBeOpen) => {
                setToggleTooltipOpen(shouldBeOpen)
              }}
              open={isToggleTooltipOpen}
            >
              <Button
                type="text"
                icon={isCollapsed ? <ExpandIcon /> : <CollapseIcon />}
                onClick={toggleCollapse}
                className={cx('toggle-button')}
              />
            </Tooltip>
          </div>
          {headerSlot && <div className={cx('mt-5')}>{headerSlot}</div>}
        </div>
        <div className={cx('flex', 'flex-col', 'justify-between', 'flex-1')}>
          <Menu
            key={`nav-menu-main-${isCollapsed}`}
            mode={'inline'}
            compact={isCollapsed}
            items={items}
            selectedKeys={[selectedKey]}
            defaultOpenKeys={isCollapsed ? undefined : defaultOpenKeys}
            onSelect={(info) => setSelectedKey(info.key)}
          />
          <Menu
            key={`nav-menu-footer-${isCollapsed}`}
            mode={'inline'}
            compact={isCollapsed}
            selectable={false}
            items={footerItems}
          />
        </div>
      </div>
    </div>
  )
}
