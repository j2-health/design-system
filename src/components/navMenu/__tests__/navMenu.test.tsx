import { act, render } from '@testing-library/react'
import { NavMenu } from '../NavMenu'

describe('NavMenu', () => {
  it('should render correctly', async () => {
    const { container } = await act(async () =>
      render(
        <NavMenu
          items={[
            { label: 'Dashboard', key: 'dashboard' },
            { label: 'Settings', key: 'settings' },
          ]}
          footerItems={[{ label: 'Logout', key: 'logout' }]}
        />
      )
    )

    expect(container).toMatchSnapshot()
  })
})
