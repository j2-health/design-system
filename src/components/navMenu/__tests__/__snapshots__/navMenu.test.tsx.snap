// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenu should render correctly 1`] = `
<div>
  <div
    class="navMenuContainer bg-white"
  >
    <div
      class="flex flex-col h-full"
    >
      <div
        class="py-5 px-3"
      >
        <div
          class="flex items-center justify-between headerContainer min-h-9"
        >
          <a
            class="flex items-center gap-2 logo"
            href="/"
          >
            <test-file-stub />
            <test-file-stub />
          </a>
          <div>
            <button
              class="ant-btn css-dev-only-do-not-override-tpassh ant-btn-text ant-btn-color-default ant-btn-variant-text ant-btn-icon-only toggle-button"
              type="button"
            >
              <span
                class="ant-btn-icon"
              >
                <test-file-stub />
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="flex flex-col justify-between flex-1"
      >
        <ul
          class="ant-menu ant-menu-root ant-menu-inline ant-menu-light j2Menu css-dev-only-do-not-override-tpassh"
          data-menu-list="true"
          role="menu"
          tabindex="0"
        >
          <li
            class="ant-menu-item ant-menu-item-only-child"
            data-menu-id="rc-menu-uuid-test-dashboard"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              class="ant-menu-title-content"
            >
              Dashboard
            </span>
          </li>
          <li
            class="ant-menu-item ant-menu-item-only-child"
            data-menu-id="rc-menu-uuid-test-settings"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              class="ant-menu-title-content"
            >
              Settings
            </span>
          </li>
        </ul>
        <div
          aria-hidden="true"
          style="display: none;"
        />
        <ul
          class="ant-menu ant-menu-root ant-menu-inline ant-menu-light j2Menu css-dev-only-do-not-override-tpassh"
          data-menu-list="true"
          role="menu"
          tabindex="0"
        >
          <li
            class="ant-menu-item ant-menu-item-only-child"
            data-menu-id="rc-menu-uuid-test-logout"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              class="ant-menu-title-content"
            >
              Logout
            </span>
          </li>
        </ul>
        <div
          aria-hidden="true"
          style="display: none;"
        />
      </div>
    </div>
  </div>
</div>
`;
